File: src/foundry/engines/llm.py
"""Structured-LLM abstraction for the intelligence engines.

The LLM-backed engines depend only on :class:`StructuredLLM` - "given a system
prompt, a user prompt and a JSON schema, return a parsed JSON object". This keeps
the engines testable with :class:`FakeStructuredLLM` (no key, no network) and
isolates every OpenAI SDK detail in :class:`OpenAIStructuredLLM`, the one place
to update if the SDK or model changes.
"""

from __future__ import annotations

import json
from typing import Any, Protocol


class LLMError(RuntimeError):
    """Raised when the LLM call or its response cannot be used."""


class StructuredLLM(Protocol):
    def generate(
        self, *, system: str, user: str, schema: dict[str, Any], schema_name: str
    ) -> dict[str, Any]: ...


class FakeStructuredLLM:
    """Test double that returns canned responses and records the prompts it saw."""

    def __init__(self, responses: list[dict[str, Any]]) -> None:
        self._responses = list(responses)
        self.calls: list[dict[str, Any]] = []

    def generate(
        self, *, system: str, user: str, schema: dict[str, Any], schema_name: str
    ) -> dict[str, Any]:
        self.calls.append({"system": system, "user": user, "schema_name": schema_name})
        if not self._responses:
            raise LLMError("FakeStructuredLLM ran out of canned responses")
        return self._responses.pop(0)


class OpenAIStructuredLLM:
    """OpenAI (GPT-5.5) structured-output backend via the Chat Completions API.

    The client is injected so this stays unit-testable; when omitted it is created
    lazily from the environment (``OPENAI_API_KEY``), which is why ``openai`` is an
    optional dependency rather than a hard one.
    """

    DEFAULT_MODEL = "gpt-5.5"

    def __init__(
        self,
        *,
        client: Any | None = None,
        model: str = DEFAULT_MODEL,
        strict: bool = False,
    ) -> None:
        self._client = client
        self._model = model
        # strict=True asks OpenAI to enforce the schema; leave off until the
        # Pydantic-derived schema has been tuned for strict mode.
        self._strict = strict

    def _ensure_client(self) -> Any:
        if self._client is None:  # pragma: no cover - requires the SDK + a key
            try:
                from openai import OpenAI
            except ImportError as exc:
                raise LLMError(
                    "openai is not installed; install the 'llm' extra to use "
                    "OpenAIStructuredLLM"
                ) from exc
            self._client = OpenAI()
        return self._client

    def generate(
        self, *, system: str, user: str, schema: dict[str, Any], schema_name: str
    ) -> dict[str, Any]:
        client = self._ensure_client()
        try:
            response = client.chat.completions.create(
                model=self._model,
                messages=[
                    {"role": "system", "content": system},
                    {"role": "user", "content": user},
                ],
                response_format={
                    "type": "json_schema",
                    "json_schema": {
                        "name": schema_name,
                        "schema": schema,
                        "strict": self._strict,
                    },
                },
            )
        except Exception as exc:
            # Callers' degrade-to-floor contracts catch LLMError only; a raw
            # RateLimitError/APITimeoutError/connection error must not leak
            # past this seam (it would abort intake or PR-event processing).
            raise LLMError(f"OpenAI API call failed: {exc}") from exc
        content = response.choices[0].message.content
        if not content:
            raise LLMError("OpenAI returned an empty response")
        try:
            return json.loads(content)
        except json.JSONDecodeError as exc:
            raise LLMError(f"OpenAI response was not valid JSON: {exc}") from exc

File: dotnet/src/Foundry/Engines/Llm.cs
// Structured-LLM abstraction for the intelligence engines.
//
// The LLM-backed engines depend only on IStructuredLlm - "given a system
// prompt, a user prompt and a schema name, return a parsed JSON object". This
// keeps the engines testable with FakeStructuredLlm (no key, no network) and
// isolates every vendor SDK detail behind one seam: implement IStructuredLlm
// over the OpenAI .NET SDK (or any other client) to go live.
//
// Note vs the Python port: Pydantic generates the JSON schema from the model;
// .NET 8 has no built-in JSON schema generation, so the schema contract here
// is the schema *name* plus strict deserialisation into the artifact type
// (unknown fields rejected, ranges validated). The governance result is the
// same: invalid model output never becomes an artifact.

using System.Text.Json;
using System.Text.Json.Nodes;
using Foundry.Schemas;

namespace Foundry.Engines;

/// <summary>Raised when the LLM call or its response cannot be used.</summary>
public class LlmException : Exception
{
    public LlmException(string message) : base(message) { }
}

public interface IStructuredLlm
{
    JsonObject Generate(string system, string user, string schemaName);
}

/// <summary>Test double that returns canned responses and records the prompts it saw.</summary>
public sealed class FakeStructuredLlm : IStructuredLlm
{
    private readonly Queue<JsonObject> _responses;

    public FakeStructuredLlm(IEnumerable<JsonObject> responses)
    {
        _responses = new Queue<JsonObject>(responses);
    }

    public List<(string System, string User, string SchemaName)> Calls { get; } = new();

    public JsonObject Generate(string system, string user, string schemaName)
    {
        Calls.Add((system, user, schemaName));
        if (_responses.Count == 0)
        {
            throw new LlmException("FakeStructuredLlm ran out of canned responses");
        }
        return _responses.Dequeue();
    }
}

/// <summary>
/// LLM-backed ticket analyzer.
///
/// Implements ITicketAnalyzer using an IStructuredLlm. This is the
/// *pre-approval gate* intelligence: judge readiness, surface what's missing,
/// and normalise acceptance criteria from natural-language tickets - the thing
/// the heuristic analyzer can't actually do. It deliberately does NOT plan the
/// implementation; that stays with the coding agent.
///
/// Robustness: the model output is validated against the TicketAnalysis schema
/// and retried with corrective feedback; identity fields are overwritten from
/// the ticket so a hallucinated id/title can't slip through.
/// </summary>
public sealed class LlmTicketAnalyzer : ITicketAnalyzer
{
    private const string SystemPrompt = """
You analyse engineering tickets. You do NOT write code and you do NOT produce an \
implementation plan. Classify the work, identify missing information, extract or \
normalise acceptance criteria, and decide whether the ticket is ready to implement.

Hard rules:
- Return ONLY a JSON object matching the TicketAnalysis schema.
- Do not invent facts. List anything you infer under "assumptions", separate from facts.
- If the ticket is unclear or lacks acceptance criteria, set implementation_readiness \
to "needs_clarification" and list what is missing.
- If the ticket is a question rather than a unit of work, set "not_suitable".
- For a bug, missing reproduction steps means it is not ready.
- Only set "ready" when the work is clear enough that a coding agent could start.
""";

    private readonly IStructuredLlm _llm;
    private readonly int _maxAttempts;

    public LlmTicketAnalyzer(IStructuredLlm llm, int maxAttempts = 2)
    {
        _llm = llm;
        _maxAttempts = Math.Max(1, maxAttempts);
    }

    public TicketAnalysis Analyse(RawTicket ticket)
    {
        var user = RenderUser(ticket);
        Exception? lastError = null;

        for (var attempt = 0; attempt < _maxAttempts; attempt++)
        {
            var prompt = attempt == 0 ? user : $"{user}\n\n{Feedback(lastError)}";
            var raw = _llm.Generate(SystemPrompt, prompt, "TicketAnalysis");
            // Identity comes from the ticket, never the model.
            raw["ticket_id"] = string.IsNullOrEmpty(ticket.IssueKey) ? ticket.IssueId : ticket.IssueKey;
            raw["title"] = ticket.Title;
            try
            {
                return FoundryJson.Deserialize<TicketAnalysis>(
                    raw.ToJsonString(new JsonSerializerOptions()));
            }
            catch (SchemaValidationException exc)
            {
                lastError = exc;
            }
        }

        throw new LlmException(
            $"LLM analyzer could not produce a valid TicketAnalysis after "
            + $"{_maxAttempts} attempts: {lastError?.Message}");
    }

    private static string RenderUser(RawTicket ticket)
    {
        var parts = new List<string>
        {
            $"Issue: {(string.IsNullOrEmpty(ticket.IssueKey) ? ticket.IssueId : ticket.IssueKey)}",
            $"Title: {ticket.Title}",
            $"Labels: {(ticket.Labels.Count > 0 ? string.Join(", ", ticket.Labels) : "(none)")}",
            "",
            "Description:",
            string.IsNullOrEmpty(ticket.Description) ? "(empty)" : ticket.Description,
        };
        if (ticket.Comments.Count > 0)
        {
            parts.Add("");
            parts.Add("Comments:");
            parts.AddRange(ticket.Comments);
        }
        return string.Join("\n", parts);
    }

    private static string Feedback(Exception? error) =>
        "Your previous response was invalid and rejected by the schema "
        + $"validator:\n{error?.Message}\nReturn a corrected JSON object only.";
}
